src/v2/auth.rs

```diff
@@ -1,23 +1,165 @@
 use crate::errors::{Error, Result};
 use crate::v2::*;
-use reqwest::{StatusCode, Url};
+use reqwest::{header::HeaderValue, RequestBuilder, StatusCode, Url};
+use std::iter::FromIterator;
+
+/// Represents all supported authentication schemes and is stored by `Client`.
+#[derive(Debug, Clone)]
+pub enum Auth {
+    Bearer(BearerAuth),
+    Basic(BasicAuth),
+}
+
+impl Auth {
+    /// Add authentication headers to a request builder.
+    pub(crate) fn add_auth_headers(&self, request_builder: RequestBuilder) -> RequestBuilder {
+        match self {
+            Auth::Bearer(bearer_auth) => request_builder.bearer_auth(bearer_auth.token.clone()),
+            Auth::Basic(basic_auth) => {
+                request_builder.basic_auth(basic_auth.user.clone(), basic_auth.password.clone())
+            }
+        }
+    }
+}
 
-#[derive(Debug, Default, Deserialize, Serialize)]
-pub struct TokenAuth {
+/// Used for Bearer HTTP Authentication
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct BearerAuth {
     token: String,
     expires_in: Option<u32>,
     issued_at: Option<String>,
     refresh_token: Option<String>,
 }
 
-impl TokenAuth {
-    pub fn token(&self) -> &str {
-        self.token.as_str()
+/// Used for Basic HTTP Authentication
+#[derive(Debug, Clone)]
+pub struct BasicAuth {
+    user: String,
+    password: Option<String>,
+}
+
+/// Structured representation for the content of the authentication response header.
+#[derive(Debug, PartialEq, Eq, Deserialize)]
+pub(crate) enum WwwAuthenticateHeaderContent {
+    Bearer(WwwAuthenticateHeaderContentBearer),
+    Basic(WwwAuthenticateHeaderContentBasic),
+}
+
+impl WwwAuthenticateHeaderContent {
+    /// Create a `WwwAuthenticateHeaderContent` by parsing a `HeaderValue` instance.
+    pub(crate) fn from_www_authentication_header(header_value: HeaderValue) -> Result<Self> {
+        let header = String::from_utf8(header_value.as_bytes().to_vec())?;
+
+        // This regex will result in multiple captures which will contain one key-value pair each.
+        // The first capture will be the only one with the "method" group set.
+        let re = regex::Regex::new(
+            r#"(?x)\s*
+            ((?P<method>[A-Z][a-z]+)\s*)?
+            (
+                \s*
+                    (?P<key>[a-z]+)
+                \s*
+                    =
+                \s*
+                    "(?P<value>[^"]+)"
+                \s*
+            )
+        "#,
+        )?;
+        let captures = re.captures_iter(&header).collect::<Vec<_>>();
+
+        let method = captures
+            .get(0)
+            .ok_or_else(|| {
+                Error::from(format!("regex '{}' didn't match '{}'", re.as_str(), header))
+            })?
+            .name("method")
+            .ok_or_else(|| Error::from(format!("method not found in {}", header)))?
+            .as_str();
+
+        let serialized_content = serde_json::Value::Object(serde_json::Map::from_iter(
```

Comment:
This is somewhat inconvenient. I'm thinking of assembling a JSON string using plain string formatting and skipping the `Value::Object` route.

Reply:
Added https://github.com/camallo/dkregistry-rs/pull/159/commits/68da0dc2f657c2566ebbba8f927268b4c11e50e3 on top which implements the plain string route. I like the aspect that we don't need the intermediate `Value` and I can't spot any weakness in the approach. I'd appreciate thoughts here.

Reply:
@lucab any opinion on this one?

Reply:
[sorry I didn't see this comment in review mode in GH]

The main problem I see with the logic in the fixup commit is when either `key` or `value` contains some characters that may need to be escaped in JSON-encoding (e.g. `"`). serde serialization takes care of that for you and is in general more robust.

But I do agree this code is a bit unwieldy. I think the source of the pain is the round-trip through JSON and serde. As a manual alternative, I think you could shove the key-value pairs into a map, drain the well-known values you expect in order to build the object, and warn if there are any leftover keys in the map.

Reply:
I found a compromise by wrapping the strings in a `serde_json::Value::String`. See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=721415923973d82b083b7e82f6d5e320). WDYT?

Reply:
I pushed this suggestion into the fixup commit and am proposing it as the final solution. It doesn't have the escaping bug you detected with the previous revision, and it doesn't do the round trip. Hence I like 0ba493ed37ee2a5d71713664e2bcae76cd2ed3c6 and I'm proposing it as the final solution. PTAL and let me know if I should squash :wink:

Reply:
Yeah, it looks better than the initial version. Feel free to squash in.